lib/adapters/parsers/grok.js

```diff
@@ -0,0 +1,71 @@
+var base = require('./base');
+var errors = require('../../errors');
+var grok = require('node-grok');
+var Promise = require('bluebird');
+var readline = require('readline');
+var sync = require('sync');
+
+module.exports = base.extend({
+
+    initialize: function(options) {
+        this.pattern = options.pattern || '%{GREEDYDATA:message}';
+    },
+
+    parseStream: function(stream, emit) {
+        var self = this;
+
+        return new Promise(function(resolve, reject) {
+            var buffer = [];
+            var pending = 0;
+            var streamEnd = false;
+
+            function checkDone(streamEnd) {
+                if (buffer.length >= self.limit) {
+                    emit(buffer);
+                    buffer = [];
+                }
+
+                if (pending === 0 && streamEnd) {
+                    emit(buffer);
+                    buffer = [];
+                    resolve();
+                }
+            }
+
+            grok.loadDefault(function(patterns) {
+                var pattern = patterns.createPattern(self.pattern);
+                var liner = readline.createInterface({
+                    input: stream,
+                    terminal: false
+                });
+
+                liner.on('line', function(line) {
+                    pending++;
+                    sync(function() {
```

Comment:
`pattern.parse.sync` isn't from the `node-grok` package (which had the asynchronous function `pattern.parse`), its added to the prototype of all functions through the [sync](https://www.npmjs.com/package/sync) library which then allows you to synchronize an asynchronous function but only when executed within the closure of `sync(function() { ... })`. I may have miss read their examples and how to use the lib so give their README a look and correct me if I'm wrong.

The only reason for this is that if I keep things using the asynchronous method its just a mess to figure out how to reorder things. I even tried promisifying the whole thing and chaining promises (to guarantee the order of that the lines are read from are maintained when emitting points) but things just get even harder to follow at that point.